### PR TITLE
[NEW UI] Update options page

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Handle update option with a brand new page
| Type?             | new feature
| BC breaks?        | Yes, the config var `PS_AUTOUP_KEEP_EMAIL` is now `PS_AUTOUP_REGEN_EMAIL`
| Deprecations?     | Nope
| Fixed ticket?     | /
| Sponsor company   | @PrestaShopCorp
| How to test?      | Update options can be updated via the new UI. Refreshing a page after an option is switched on or off must keep its value.

image

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
Requires the 2 following PRs to be merged:
* https://github.com/PrestaShop/autoupgrade/pull/975
* https://github.com/PrestaShop/autoupgrade/pull/980

File: views/templates/steps/update-options.html.twig
Comment (Quetzacoalt91) [inline views/templates/steps/update-options.html.twig]:
```suggestion
      title: 'Deactivate non-native modules'|trans({}),
```

File: _dev/src/ts/pages/UpdatePageUpdateOptions.ts
Comment (Quetzacoalt91) [inline _dev/src/ts/pages/UpdatePageUpdateOptions.ts]: @ga-devfront I've kept this implementation after trying switching on the whole form in FormData. Unfortunately the options are only sent if they are checked, which does not fit the ConfigurationValidator behavior where only updated data is expected.

File: README.md
Comment (M0rgan01) [inline README.md:105]: the comments were minor but helped understand the configurations. Maybe adding an explanation on the README or a link to the devdoc would help

File: classes/Analytics.php
Comment (M0rgan01) [inline classes/Analytics.php]: maybe we could also change the tag on segment? We could manage this by version right?

File: controllers/admin/self-managed/UpdatePageUpdateOptionsController.php
Comment (M0rgan01) [inline controllers/admin/self-managed/UpdatePageUpdateOptionsController.php]: We could wait for the PR https://github.com/PrestaShop/autoupgrade/pull/993 and thus use the new constants, same thing in getParams

File: controllers/admin/self-managed/UpdatePageUpdateOptionsController.php
Comment (Quetzacoalt91) [inline controllers/admin/self-managed/UpdatePageUpdateOptionsController.php]: Rename as `$errors`

Comment (Quetzacoalt91) [inline controllers/admin/self-managed/UpdatePageUpdateOptionsController.php]: Data structure is incorrect for template and needs to be reworked before being sent. Reuse logic from `controllers/admin/self-managed/UpdatePageVersionChoiceController.php::save()`

File: _dev/src/ts/pages/UpdatePageUpdateOptions.ts
Comment (ga-devfront) [reply _dev/src/ts/pages/UpdatePageUpdateOptions.ts]: you can do it like this to get all the form with all chebox true or false =>
```
const checkboxStates = Array.from(this.form.elements)
  .filter(el => el.type === 'checkbox')
  .reduce((acc, checkbox) => {
    acc[checkbox.name] = checkbox.checked;
    return acc;
  }, {});
  ```

File: controllers/admin/self-managed/UpdatePageUpdateOptionsController.php
Comment (ga-devfront) [reply controllers/admin/self-managed/UpdatePageUpdateOptionsController.php]: +1

File: classes/Analytics.php
Comment (Quetzacoalt91) [reply classes/Analytics.php]: I needed to check with the data tcheam first, but yeah we can do that knowing the data will be reset on the date of the v7 release.

File: README.md
Comment (Quetzacoalt91) [reply README.md:105]: Good idea, we can go further than a unique example.
I've created a task for this matter so it can go in a dedicated PR: SUE-248

**Review by M0rgan01 — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
Blocked until we get an additional acceptance criteria from Geric.

File: classes/Parameters/ConfigurationValidator.php
Comment (M0rgan01) [inline classes/Parameters/ConfigurationValidator.php:116]: I don't think it's a good practice to add mixed typing, especially if one day we want to upgrade phpstan level. This function was intended to validate the boolean in the form of a string, perhaps we could do this line 71, or in another function

File: classes/Parameters/UpgradeConfiguration.php
Comment (M0rgan01) [inline classes/Parameters/UpgradeConfiguration.php]: If the input array contains something other than the accepted configurations, then an error should probably be raised, or the element in question should be logged.

File: classes/Parameters/ConfigurationValidator.php
Comment (Quetzacoalt91) [reply classes/Parameters/ConfigurationValidator.php:116]: Let's remember these methods return an error when something is wrong. If we go with another method, this new one would return a boolean to let us know if further checks are needed or not. This makes the contents of the class more confusing.

Having this in the switch / case list adds unnecessary complexity as well. We have calls to a single method for each case and I like the way it currently is. Simple and efficient.

The `mixed` case does not apply here as we specify two types in the php doc block. Even if the freshly introduce lvl 10 of PHPStan, this definition pass:
image

File: classes/Parameters/UpgradeConfiguration.php
Comment (Quetzacoalt91) [reply classes/Parameters/UpgradeConfiguration.php]: I've reverted this change so we can move forward.

File: classes/Parameters/ConfigurationValidator.php
Comment (M0rgan01) [reply classes/Parameters/ConfigurationValidator.php:116]: > Having this in the switch / case list adds unnecessary complexity as well.

I don't see why this would add complexity, we would just have the check before calling the method, which could be renamed to explain that we are validating a boolean "string"

PHPstan may not say anything, but this remains a practice that does not follow modern PHP practices. This would not have been possible on languages ​​with (really) strong typing, and I think it's better.

I think that on the contrary, a parameter with multiple types can make the function more difficult to understand, because it can be difficult to know how the function will behave depending on the type passed.

Comment (M0rgan01) [reply classes/Parameters/ConfigurationValidator.php:116]: But I don't want to block the PR for this detail, we have to move it forward :+1:

**Review by ga-devfront — DISMISSED** [review]
image

**Comment by Quetzacoalt91** [comment]
Following latest conversation with @gericfo:
* The current behavior on the switches is acceptable in case of errors,
* Some acceptance criteria will be added on the related user story, and will be done on another pull-request.

**Comment by Quetzacoalt91** [comment]
Had to rebase after the merge of the latest pull-requests

Here we go again

**Review by ga-devfront — APPROVED** [review]

**Review by AureRita — APPROVED** [review]
Hi @Quetzacoalt91 

Thank you for your PR, I tested it and it seems to works as you can see : 

[recording(333).webm](https://github.com/user-attachments/assets/2a10132d-7549-4079-89eb-a7f58eeceb25)


Tested from : 
1.7.8.8 to 8.2 
8.0.4 to 8.2 
8.2 to 9.0.0
8.0.4 to 9.0.0

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you

**Pull request merged**
